gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -407,18 +423,65 @@ private long getCommittedWriteSize(String uploadId) throws InterruptedException,
 
       SimpleResponseObserver<QueryWriteStatusResponse> responseObserver =
           new SimpleResponseObserver<>();
+      GrpcQueryWriteStatusRequestExecutor requestExecutor =
+          new GrpcQueryWriteStatusRequestExecutor<>(request, responseObserver);
+      try {
+        ResilientOperation.retry(
+            requestExecutor,
+            backOffFactory.newBackOff(),
+            rateLimitedRetryDeterminer,
+            IOException.class,
+            sleeper);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new IOException("Failed to get committed write size.", e);
+      }
 
-      stub.withDeadlineAfter(QUERY_WRITE_STATUS_TIMEOUT.toMillis(), TimeUnit.MILLISECONDS)
-          .queryWriteStatus(request, responseObserver);
-      responseObserver.done.await();
+      return responseObserver.getResponse().getCommittedSize();
+    }
 
-      if (responseObserver.hasError()) {
-        throw new IOException(
-            String.format("Failed to get committed write size '%s'", resourceId),
-            responseObserver.getError());
+    /** Simple class to create a {@link CheckedCallable} from a {@link QueryWriteStatusRequest}. */
+    private class GrpcQueryWriteStatusRequestExecutor<T>
```

Comment:
Can you use lambda instead of creating custom class to execute retries?

Reply:
Doubt I can do so, the retry method requires a CheckedCallable to be passed as a wrapper of the code to be executed. Secondly CheckedCallable actually provide a good generic for exceptions to be thrown, which is a little bit messy to do with lambda.